server/src/main/java/org/apache/druid/server/coordinator/SegmentStateManager.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.server.coordinator;
+
+import com.google.inject.Inject;
+import org.apache.druid.client.ServerInventoryView;
+import org.apache.druid.java.util.emitter.EmittingLogger;
+import org.apache.druid.metadata.SegmentsMetadataManager;
+import org.apache.druid.timeline.DataSegment;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Manages state of segments being loaded.
+ */
+public class SegmentStateManager
+{
+  private static final EmittingLogger log = new EmittingLogger(SegmentStateManager.class);
+
+  private final LoadQueueTaskMaster taskMaster;
+  private final ServerInventoryView serverInventoryView;
+  private final SegmentsMetadataManager segmentsMetadataManager;
+
+  private final ConcurrentHashMap<String, TierLoadingState> currentlyMovingSegments =
+      new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, TierLoadingState> currentlyReplicatingSegments
+      = new ConcurrentHashMap<>();
+
+  @Inject
+  public SegmentStateManager(
+      ServerInventoryView serverInventoryView,
+      SegmentsMetadataManager segmentsMetadataManager,
+      LoadQueueTaskMaster taskMaster
+  )
+  {
+    this.serverInventoryView = serverInventoryView;
+    this.segmentsMetadataManager = segmentsMetadataManager;
+    this.taskMaster = taskMaster;
+  }
+
+  /**
+   * Queues load of a replica of the segment on the given server.
+   */
+  public boolean loadSegment(
+      DataSegment segment,
+      ServerHolder server,
+      boolean isPrimary,
+      ReplicationThrottler throttler
+  )
+  {
+    final String tier = server.getServer().getTier();
+    final LoadPeonCallback callback;
+    if (isPrimary) {
+      // Primary replicas are not subject to throttling
+      callback = null;
```

Comment:
What's the definition of "isPrimary"?  There really isn't a notion of "primary" or "secondary" replication, at least there shouldn't be.  The only semantic difference that should exist is

1) If the data is not available at all inside of the tier -> get the data loaded as fast as humanly possibly
2) If the data is available somewhere inside of the tier -> throttle the loading so that there is capacity in the cluster for other activities as well as under the assumption that this usually happens due to a failure and the node could just be flapping.

Reply:
Yes, I have made this change in a follow up PR which fully enables prioritisation and does away with the notion of primary and secondary. Should I include that here itself?

Reply:
Fixed.